File: firebase/dart_functions/lib/secret_manager.dart
import 'dart:convert';
import 'dart:io';

import 'package:google_cloud_secretmanager_v1/secretmanager.dart';
import 'package:googleapis_auth/auth_io.dart';

String resolveProjectId([String? override]) {
  final projectId = override ?? Platform.environment['GCLOUD_PROJECT'];
  if (projectId == null || projectId.trim().isEmpty) {
    throw Exception('GCLOUD_PROJECT environment variable is not set.');
  }
  return projectId;
}

String buildSecretPath(String projectId, String secretId) {
  if (secretId.trim().isEmpty) {
    throw Exception('secretId must not be empty.');
  }
  return 'projects/$projectId/secrets/$secretId/versions/latest';
}

String extractSecretData(
  AccessSecretVersionResponse response,
  String secretId,
) {
  final data = response.payload?.data;
  if (data == null) {
    throw Exception('Secret "$secretId" has no data.');
  }
  return utf8.decode(data);
}

Future<SecretManagerService> _createClient() async {
  final httpClient = await clientViaMetadataServer();
  return SecretManagerService(client: httpClient);
}

Future<String> accessSecret(String secretId) async {
  final projectId = resolveProjectId();
  final client = await _createClient();
  try {
    final response = await client.accessSecretVersion(
      AccessSecretVersionRequest(name: buildSecretPath(projectId, secretId)),
    );
    return extractSecretData(response, secretId);
  } finally {
    client.close();
  }
}

Future<String> createSecretWithValue(String secretId, String value) async {
  final projectId = resolveProjectId();
  final parent = 'projects/$projectId';
  final client = await _createClient();
  try {
    await client.createSecret(
      CreateSecretRequest(
        parent: parent,
        secretId: secretId,
        secret: Secret(
          replication: Replication(automatic: Replication_Automatic()),
        ),
      ),
    );

    await client.addSecretVersion(
      AddSecretVersionRequest(
        parent: '$parent/secrets/$secretId',
        payload: SecretPayload(data: utf8.encode(value)),
      ),
    );

    return '$parent/secrets/$secretId';
  } finally {
    client.close();
  }
}

Future<void> addSecretVersionByPath(String secretPath, String value) async {
  final client = await _createClient();
  try {
    await client.addSecretVersion(
      AddSecretVersionRequest(
        parent: secretPath,
        payload: SecretPayload(data: utf8.encode(value)),
      ),
    );
  } finally {
    client.close();
  }
}

Future<void> deleteSecretByPath(String secretPath) async {
  final client = await _createClient();
  try {
    await client.deleteSecret(DeleteSecretRequest(name: secretPath));
  } finally {
    client.close();
  }
}

File: firebase/ts_functions/src/secretManager.ts
import { SecretManagerServiceClient } from "@google-cloud/secret-manager";

interface AccessSecretVersionResponse {
  payload?: {
    data?: Uint8Array | string | null;
  } | null;
}

export function resolveProjectId(override?: string): string {
  const projectId = override ?? process.env.GCLOUD_PROJECT ?? process.env.GOOGLE_CLOUD_PROJECT;
  if (!projectId || projectId.trim().length === 0) {
    throw new Error("GCLOUD_PROJECT environment variable is not set.");
  }
  return projectId;
}

export function buildSecretPath(projectId: string, secretId: string): string {
  if (secretId.trim().length === 0) {
    throw new Error("secretId must not be empty.");
  }
  return `projects/${projectId}/secrets/${secretId}/versions/latest`;
}

export function extractSecretData(
  response: AccessSecretVersionResponse,
  secretId: string,
): string {
  const data = response.payload?.data;
  if (!data) {
    throw new Error(`Secret "${secretId}" has no data.`);
  }
  return Buffer.from(data).toString("utf8");
}

export async function accessSecret(secretId: string): Promise<string> {
  const projectId = resolveProjectId();
  const client = new SecretManagerServiceClient();
  const [response] = await client.accessSecretVersion({
    name: buildSecretPath(projectId, secretId),
  });
  return extractSecretData(response, secretId);
}

export async function createSecretWithValue(secretId: string, value: string): Promise<string> {
  const projectId = resolveProjectId();
  const parent = `projects/${projectId}`;
  const client = new SecretManagerServiceClient();

  await client.createSecret({
    parent,
    secretId,
    secret: {
      replication: {
        automatic: {},
      },
    },
  });

  await client.addSecretVersion({
    parent: `${parent}/secrets/${secretId}`,
    payload: {
      data: Buffer.from(value, "utf8"),
    },
  });

  return `${parent}/secrets/${secretId}`;
}

export async function addSecretVersionByPath(secretPath: string, value: string): Promise<void> {
  const client = new SecretManagerServiceClient();
  await client.addSecretVersion({
    parent: secretPath,
    payload: {
      data: Buffer.from(value, "utf8"),
    },
  });
}

export async function deleteSecretByPath(secretPath: string): Promise<void> {
  const client = new SecretManagerServiceClient();
  await client.deleteSecret({ name: secretPath });
}
